File: python/x402/mechanisms/evm/multicall.py
"""Helpers for batching `eth_call` requests through Multicall3."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any

try:
    from eth_abi import decode, encode
    from eth_utils import keccak
except ImportError as e:
    raise ImportError(
        "EVM mechanism requires ethereum packages. Install with: pip install x402[evm]"
    ) from e

from .constants import MULTICALL3_ADDRESS, MULTICALL3_TRY_AGGREGATE_ABI
from .signer import FacilitatorEvmSigner


@dataclass
class MulticallCall:
    """One call executed through Multicall3."""

    address: str
    abi: list[dict[str, Any]] | None = None
    function_name: str = ""
    args: tuple[Any, ...] = field(default_factory=tuple)
    call_data: bytes = b""


@dataclass
class MulticallResult:
    """Decoded result for a single multicall entry."""

    success: bool
    result: Any = None
    error: Exception | None = None


def encode_contract_call(
    abi: list[dict[str, Any]],
    function_name: str,
    *args: Any,
) -> bytes:
    """Encode calldata for a contract function."""
    function = _get_function_abi(abi, function_name)
    input_types = [_canonical_type(item) for item in function.get("inputs", [])]
    signature = f"{function_name}({','.join(input_types)})"
    selector = keccak(text=signature)[:4]
    return selector + encode(input_types, list(args))


def multicall(
    signer: FacilitatorEvmSigner,
    calls: list[MulticallCall],
) -> list[MulticallResult]:
    """Batch calls through Multicall3 and decode the results."""
    if not calls:
        return []

    aggregate_calls = []
    for call in calls:
        call_data = call.call_data
        if not call_data:
            if not call.abi or not call.function_name:
                raise ValueError("typed multicall entries require ABI and function name")
            call_data = encode_contract_call(call.abi, call.function_name, *call.args)
        aggregate_calls.append((call.address, call_data))

    raw_results = signer.read_contract(
        MULTICALL3_ADDRESS,
        MULTICALL3_TRY_AGGREGATE_ABI,
        "tryAggregate",
        False,
        aggregate_calls,
    )
    normalized = _normalize_results(raw_results)

    if len(normalized) != len(calls):
        raise ValueError(
            f"multicall result length mismatch: got {len(normalized)}, want {len(calls)}"
        )

    results: list[MulticallResult] = []
    for raw_result, call in zip(normalized, calls, strict=True):
        success, return_data = raw_result
        if not success:
            results.append(
                MulticallResult(
                    success=False,
                    error=RuntimeError("multicall: call reverted"),
                )
            )
            continue

        if call.call_data:
            results.append(MulticallResult(success=True))
            continue

        try:
            decoded = _decode_contract_result(call.abi or [], call.function_name, return_data)
        except Exception as exc:
            results.append(MulticallResult(success=False, error=exc))
            continue

        results.append(MulticallResult(success=True, result=decoded))

    return results


def _decode_contract_result(
    abi: list[dict[str, Any]],
    function_name: str,
    return_data: bytes,
) -> Any:
    function = _get_function_abi(abi, function_name)
    output_types = [_canonical_type(item) for item in function.get("outputs", [])]
    if not output_types:
        return None

    decoded = decode(output_types, return_data)
    if len(decoded) == 1:
        return decoded[0]
    return list(decoded)


def _get_function_abi(abi: list[dict[str, Any]], function_name: str) -> dict[str, Any]:
    for entry in abi:
        if entry.get("type") == "function" and entry.get("name") == function_name:
            return entry
    raise ValueError(f"Function {function_name} not found in ABI")


def _canonical_type(abi_item: dict[str, Any]) -> str:
    item_type = abi_item["type"]
    if not item_type.startswith("tuple"):
        return item_type

    suffix = item_type[len("tuple") :]
    components = ",".join(_canonical_type(component) for component in abi_item["components"])
    return f"({components}){suffix}"


def _normalize_results(raw_results: Any) -> list[tuple[bool, bytes]]:
    if not isinstance(raw_results, list | tuple):
        raise ValueError(f"multicall returned {type(raw_results)!r}, want sequence")

    normalized: list[tuple[bool, bytes]] = []
    for index, entry in enumerate(raw_results):
        if isinstance(entry, dict):
            success = bool(entry["success"])
            return_data = entry["returnData"]
        else:
            if hasattr(entry, "success") and hasattr(entry, "returnData"):
                success = bool(entry.success)
                return_data = entry.returnData
            elif isinstance(entry, list | tuple) and len(entry) == 2:
                success = bool(entry[0])
                return_data = entry[1]
            else:
                raise ValueError(f"multicall entry {index} has unexpected type {type(entry)!r}")

        if isinstance(return_data, str):
            return_data = bytes.fromhex(return_data.removeprefix("0x"))
        if not isinstance(return_data, bytes):
            raise ValueError(
                f"multicall entry {index} returnData has unexpected type {type(return_data)!r}"
            )

        normalized.append((success, return_data))

    return normalized

File: go/mechanisms/evm/multicall.go
package evm

import (
	"context"
	"fmt"
	"reflect"
	"strings"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
)

// MulticallCall describes one batched call.
// Use CallData for pre-encoded raw calls, or ABI/FunctionName/Args for typed calls.
type MulticallCall struct {
	Address      string
	ABI          []byte
	FunctionName string
	Args         []interface{}
	CallData     []byte
}

// MulticallResult is the decoded outcome of one batched call.
type MulticallResult struct {
	Status string
	Result interface{}
	Error  error
}

type multicallTryAggregateCall struct {
	Target   common.Address
	CallData []byte
}

type multicallTryAggregateResult struct {
	Success    bool
	ReturnData []byte
}

// Success reports whether the call completed successfully.
func (r MulticallResult) Success() bool {
	return r.Status == "success"
}

// Multicall batches typed and raw eth_call requests via Multicall3.
func Multicall(
	ctx context.Context,
	signer FacilitatorEvmSigner,
	calls []MulticallCall,
) ([]MulticallResult, error) {
	if len(calls) == 0 {
		return nil, nil
	}

	aggregateCalls := make([]multicallTryAggregateCall, 0, len(calls))
	for _, call := range calls {
		callData, err := multicallCallData(call)
		if err != nil {
			return nil, err
		}

		aggregateCalls = append(aggregateCalls, multicallTryAggregateCall{
			Target:   common.HexToAddress(call.Address),
			CallData: callData,
		})
	}

	rawResults, err := signer.ReadContract(
		ctx,
		MULTICALL3Address,
		Multicall3TryAggregateABI,
		FunctionTryAggregate,
		false,
		aggregateCalls,
	)
	if err != nil {
		return nil, err
	}

	decodedResults, err := normalizeMulticallResults(rawResults)
	if err != nil {
		return nil, err
	}
	if len(decodedResults) != len(calls) {
		return nil, fmt.Errorf("multicall result length mismatch: got %d, want %d", len(decodedResults), len(calls))
	}

	results := make([]MulticallResult, 0, len(calls))
	for i, raw := range decodedResults {
		if !raw.Success {
			results = append(results, MulticallResult{
				Status: "failure",
				Error:  fmt.Errorf("multicall: call reverted"),
			})
			continue
		}

		call := calls[i]
		if len(call.CallData) > 0 {
			results = append(results, MulticallResult{Status: "success"})
			continue
		}

		decoded, err := decodeMulticallResult(call, raw.ReturnData)
		if err != nil {
			results = append(results, MulticallResult{
				Status: "failure",
				Error:  err,
			})
			continue
		}

		results = append(results, MulticallResult{
			Status: "success",
			Result: decoded,
		})
	}

	return results, nil
}

func multicallCallData(call MulticallCall) ([]byte, error) {
	if len(call.CallData) > 0 {
		return call.CallData, nil
	}

	if len(call.ABI) == 0 || call.FunctionName == "" {
		return nil, fmt.Errorf("multicall typed call requires ABI and function name")
	}

	contractABI, err := abi.JSON(strings.NewReader(string(call.ABI)))
	if err != nil {
		return nil, fmt.Errorf("failed to parse ABI for multicall: %w", err)
	}

	data, err := contractABI.Pack(call.FunctionName, call.Args...)
	if err != nil {
		return nil, fmt.Errorf("failed to pack multicall input for %s: %w", call.FunctionName, err)
	}

	return data, nil
}

func decodeMulticallResult(call MulticallCall, returnData []byte) (interface{}, error) {
	contractABI, err := abi.JSON(strings.NewReader(string(call.ABI)))
	if err != nil {
		return nil, fmt.Errorf("failed to parse ABI for multicall decode: %w", err)
	}

	outputs, err := contractABI.Unpack(call.FunctionName, returnData)
	if err != nil {
		return nil, fmt.Errorf("failed to decode multicall result for %s: %w", call.FunctionName, err)
	}

	if len(outputs) == 0 {
		return nil, nil
	}
	if len(outputs) == 1 {
		return outputs[0], nil
	}

	return outputs, nil
}

func normalizeMulticallResults(raw interface{}) ([]multicallTryAggregateResult, error) {
	value := reflect.ValueOf(raw)
	if !value.IsValid() {
		return nil, fmt.Errorf("multicall returned nil results")
	}
	if value.Kind() != reflect.Slice {
		return nil, fmt.Errorf("multicall returned %T, want slice", raw)
	}

	results := make([]multicallTryAggregateResult, 0, value.Len())
	for i := 0; i < value.Len(); i++ {
		entry := value.Index(i)
		if entry.Kind() == reflect.Interface || entry.Kind() == reflect.Pointer {
			entry = entry.Elem()
		}
		if !entry.IsValid() || entry.Kind() != reflect.Struct {
			return nil, fmt.Errorf("multicall entry %d has unexpected type %s", i, entry.Kind())
		}

		successField := entry.FieldByName("Success")
		returnDataField := entry.FieldByName("ReturnData")
		if !successField.IsValid() || !returnDataField.IsValid() {
			return nil, fmt.Errorf("multicall entry %d missing expected fields", i)
		}

		returnData, ok := returnDataField.Interface().([]byte)
		if !ok {
			return nil, fmt.Errorf("multicall entry %d returnData has unexpected type %T", i, returnDataField.Interface())
		}

		results = append(results, multicallTryAggregateResult{
			Success:    successField.Bool(),
			ReturnData: returnData,
		})
	}

	return results, nil
}
